File: examples/frida_capture.py
# Frida 抓取编辑器调试 launcher（version-13/SCE）的 ws2_32 收发字节
# 用途：逆向 DebugManager 控制协议（EditorLogin/UploadDelta/EditorStartGame）的真实线上格式
# 用法：python frida_capture.py <out.jsonl>
import sys, json, time, frida

SCE = r"D:\sce_online\version-13\SCE"
ARGS = [
    "-server=editor-pd.spark.xd.com", "-use_local_res", "-launcher=星火编辑器.exe",
    "-editor_api_version=13", "-no_ask_editor_api_version", "-generate_and_debug_map",
    r"-file_path=c:\Users\woaye\Documents\SCE Projects\test_res002\project.sce",
]

JS = r"""
let hooked = false;
function hex(buf, len) {
    if (len <= 0) return "";
    const n = Math.min(len, 16 * 1024 * 1024);
    try { return Array.from(new Uint8Array(buf.readByteArray(n))).map(b=>b.toString(16).padStart(2,'0')).join(''); }
    catch(e) { return "<err:" + e + ">"; }
}
function parseSockaddr(p) {
    try {
        const fam = p.readU16();
        if (fam === 2) {
            const port = ((p.add(2).readU8() << 8) | p.add(3).readU8());
            const ip = [p.add(4).readU8(), p.add(5).readU8(), p.add(6).readU8(), p.add(7).readU8()].join('.');
            return ip + ':' + port;
        }
        return 'fam=' + fam;
    } catch(e) { return '?'; }
}
const sockAddr = {};
function emit(rec) { send({ t: 'io', rec: rec }); }

function installHooks() {
    if (hooked) return;
    let ws;
    try { ws = Process.getModuleByName('ws2_32.dll'); } catch(e) { setTimeout(installHooks, 500); return; }
    hooked = true;
    emit({op:'hooks_installed'});
    Interceptor.attach(ws.findExportByName('connect'), {
        onEnter(args) { this.s = args[0].toString(); this.addr = parseSockaddr(args[1]); },
        onLeave(rv) { sockAddr[this.s] = this.addr; emit({op:'connect', sock:this.s, addr:this.addr, ret:rv.toInt32()}); }
    });
    Interceptor.attach(ws.findExportByName('send'), {
        onEnter(args) {
            const s = args[0].toString(); const len = args[2].toInt32();
            emit({op:'send', sock:s, addr:sockAddr[s]||'?', len:len, data:hex(args[1], len)});
        }
    });
    Interceptor.attach(ws.findExportByName('recv'), {
        onEnter(args) { this.s = args[0].toString(); this.buf = args[1]; this.len = args[2].toInt32(); },
        onLeave(rv) {
            const n = rv.toInt32();
            if (n > 0) emit({op:'recv', sock:this.s, addr:sockAddr[this.s]||'?', len:n, data:hex(this.buf, n)});
        }
    });
    Interceptor.attach(ws.findExportByName('WSASend'), {
        onEnter(args) {
            const s = args[0].toString();
            const cnt = args[2].toInt32();
            const bufs = args[1];
            for (let i = 0; i < cnt; i++) {
                const b = bufs.add(i * Process.pointerSize * 2);
                const len = b.readU32();
                const ptr = b.add(Process.pointerSize).readPointer();
                emit({op:'WSASend', sock:s, addr:sockAddr[s]||'?', len:len, data:hex(ptr, len)});
            }
        }
    });
    Interceptor.attach(ws.findExportByName('WSARecv'), {
        onEnter(args) { this.s = args[0].toString(); this.bufs = args[1]; this.cnt = args[2].toInt32(); this.got = args[3]; },
        onLeave(rv) {
            try {
                const n = this.got.readU32();
                if (n > 0) {
                    const b = this.bufs;
                    const len = b.readU32();
                    const ptr = b.add(Process.pointerSize).readPointer();
                    emit({op:'WSARecv', sock:this.s, addr:sockAddr[this.s]||'?', len:Math.min(len,n), data:hex(ptr, Math.min(len,n))});
                }
            } catch(e) {}
        }
    });
    Interceptor.attach(ws.findExportByName('closesocket'), {
        onEnter(args) { emit({op:'close', sock:args[0].toString()}); }
    });
}
installHooks();
"""

def main():
    out_path = sys.argv[1]
    device = frida.get_local_device()
    pid = device.spawn([SCE] + ARGS, cwd=r"D:\sce_online")
    print(f"spawned pid {pid}")
    session = device.attach(pid)
    script = session.create_script(JS)
    n = [0]
    with open(out_path, 'a', encoding='utf-8') as f:
        def on_message(msg, data):
            if msg.get('type') == 'send':
                rec = msg['payload']['rec']
                n[0] += 1
                f.write(json.dumps(rec, ensure_ascii=False) + '\n')
                f.flush()
                print(f"[{n[0]}] {rec['op']} {rec.get('addr','')} len={rec.get('len','')}", flush=True)
            else:
                print("MSG:", msg, flush=True)
        script.on('message', on_message)
        script.load()
        device.resume(pid)
        print("resumed, capturing 240s...", flush=True)
        time.sleep(240)
    print("done")

if __name__ == '__main__':
    main()

File: examples/frida_capture.rs
//! Frida 抓取编辑器调试 launcher（version-13/SCE）的 ws2_32 收发字节。
//! （frida_capture.py 的 Rust 版；注入脚本为 examples/ws_hook.js，经 include_str! 嵌入）
//!
//! 用途：逆向 DebugManager 控制协议（EditorLogin/上传/EditorStartGame）的真实线上格式。
//!
//! 用法：
//!   cargo run --release --example frida_capture -- <out.jsonl> [选项] [-- <目标进程参数>...]
//! 选项：
//!   --exe <路径>     目标程序（默认 D:/sce_online/version-13/SCE）
//!   --cwd <目录>     工作目录（默认 D:/sce_online）
//!   --seconds <秒>   抓包时长（默认 240）
//!
//! 默认目标参数 = 编辑器「生成并调试」test_res002 项目的 launcher 命令行。

use frida::{DeviceManager, Frida, Message, ScriptHandler, ScriptOption, SpawnOptions};
use std::io::Write;

const JS: &str = include_str!("ws_hook.js");

struct Handler {
    out: std::fs::File,
    n: usize,
}

impl ScriptHandler for Handler {
    fn on_message(&mut self, message: Message, _data: Option<Vec<u8>>) {
        // 自定义 send() 的 payload 不匹配 crate 内置的 SendPayload（RPC 形状），
        // 反序列化失败会落入 Message::Other({"error","data": 原始json串})，在此还原。
        let raw = match &message {
            Message::Other(v) => v.get("data").and_then(|d| d.as_str()).map(String::from),
            Message::Error(e) => {
                eprintln!("[script error] {} ({}:{})", e.description, e.line_number, e.column_number);
                None
            }
            Message::Log(l) => {
                eprintln!("[script log] {}", l.payload);
                None
            }
            other => {
                eprintln!("[msg] {other:?}");
                None
            }
        };
        let Some(raw) = raw else { return };
        let Ok(v) = serde_json::from_str::<serde_json::Value>(&raw) else { return };
        let Some(rec) = v.get("payload").and_then(|p| p.get("rec")) else { return };
        self.n += 1;
        let line = serde_json::to_string(rec).unwrap();
        let _ = writeln!(self.out, "{line}");
        let _ = self.out.flush();
        println!(
            "[{}] {} {} len={}",
            self.n,
            rec["op"].as_str().unwrap_or(""),
            rec["addr"].as_str().unwrap_or(""),
            rec["len"].as_i64().map(|x| x.to_string()).unwrap_or_default()
        );
    }
}

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let mut out_path: Option<String> = None;
    let mut exe = r"D:/sce_online/version-13/SCE".to_string();
    let mut cwd = r"D:/sce_online".to_string();
    let mut seconds: u64 = 240;
    let mut targs: Vec<String> = Vec::new();
    let mut it = args.iter().peekable();
    while let Some(a) = it.next() {
        match a.as_str() {
            "--exe" => exe = it.next().expect("--exe 缺值").clone(),
            "--cwd" => cwd = it.next().expect("--cwd 缺值").clone(),
            "--seconds" => seconds = it.next().expect("--seconds 缺值").parse().expect("秒数非法"),
            "--" => targs.extend(it.by_ref().cloned()),
            s if !s.starts_with('-') && out_path.is_none() => out_path = Some(s.to_string()),
            _ => {
                eprintln!("用法: frida_capture <out.jsonl> [--exe 路径] [--cwd 目录] [--seconds 秒] [-- 目标参数...]");
                std::process::exit(1);
            }
        }
    }
    let Some(out_path) = out_path else {
        eprintln!("缺输出文件路径");
        std::process::exit(1);
    };
    if targs.is_empty() {
        targs = [
            "-server=editor-pd.spark.xd.com",
            "-use_local_res",
            "-launcher=星火编辑器.exe",
            "-editor_api_version=13",
            "-no_ask_editor_api_version",
            "-generate_and_debug_map",
            r"-file_path=c:\Users\woaye\Documents\SCE Projects\test_res002\project.sce",
        ]
        .iter()
        .map(|s| s.to_string())
        .collect();
    }

    let frida = unsafe { Frida::obtain() };
    let device_manager = DeviceManager::obtain(&frida);
    let mut device = device_manager.get_local_device().expect("获取本机设备失败");

    let argv: Vec<String> = std::iter::once(exe.clone()).chain(targs.iter().cloned()).collect();
    let cwd_c = std::ffi::CString::new(cwd).unwrap();
    let options = SpawnOptions::new().argv(argv).cwd(cwd_c.as_c_str());
    let pid = device.spawn(&exe, &options).expect("spawn 失败");
    println!("spawned pid {pid}");

    let session = device.attach(pid).expect("attach 失败");
    let mut script = session
        .create_script(JS, &mut ScriptOption::default())
        .expect("create_script 失败");
    let out = std::fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(&out_path)
        .expect("打开输出文件失败");
    script
        .handle_message(Handler { out, n: 0 })
        .expect("handle_message 失败");
    script.load().expect("script load 失败");
    device.resume(pid).expect("resume 失败");
    println!("resumed, capturing {seconds}s...");
    std::thread::sleep(std::time::Duration::from_secs(seconds));
    drop(script);
    drop(session);
    let _ = device.kill(pid);
    println!("done");
}
